astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -23,212 +147,509 @@ class NDArithmeticMixin(object):
+
+        uncertainty_correlation : ``Number`` or `~numpy.ndarray`, optional
+            see :meth:`NDArithmeticMixin.add`
 
-        propagate_uncertainties : bool
-            Whether to propagate uncertainties following the propagation rules
-            defined by the class used for the ``uncertainty`` attribute.
+        kwargs :
+            Any other parameter that should be passed to the
+            different :meth:`NDArithmeticMixin._arithmetic_mask` (or wcs, ...)
+            methods.
 
         Returns
         -------
-        result : `~astropy.nddata.NDData`
-            The resulting dataset
-
-        Notes
-        -----
-        This method requires the datasets to have identical WCS
-        properties, equivalent units, and identical shapes.
-        Meta-data get set to None in the resulting dataset. The unit
-        in the result is the same as the unit in ``self``. Uncertainties
-        are propagated, although correlated errors are not supported
-        by any of the built-in uncertainty classes.  If uncertainties
-        are assumed to be correlated, a warning is issued by default
-        (though this can be disabled via the
-        ``astropy.nddata.conf.warn_unsupported_correlated``
-        configuration item). Values masked in either dataset before
-        the operation are masked in the resulting dataset.
+        result : `~numpy.ndarray` or `~astropy.units.Quantity`
+            The resulting data as array (in case both operands were without
+            unit) or as quantity if at least one had a unit.
+
+        kwargs : `dict`
+            the kwargs should contain all the other attributes (besides data
+            and unit) to create a new instance for the result. Creating the
+            new instance is up to the calling method, for example
+            :meth:`NDArithmeticMixin.add`.
+
+            TODO: Reword this? (embray)
+            I did formulate a new text, does this look better to you?
+
         """
+        # Convert the operand to the same class this allows for arithmetic
+        # operations with numbers, lists, numpy arrays, numpy masked arrays
+        # astropy quantities, masked quantities and of other subclasses of
+        # NDData
+        operand = self.__class__(operand)
+
+        # Find the appropriate keywords for the appropriate method (not sure
+        # if data and uncertainty are ever used ...)
+        # TODO: Ask about data and uncertainty :-)
+        kwds2 = {'mask': {}, 'meta': {}, 'wcs': {},
+                 'data': {}, 'uncertainty': {}}
+        for i in kwds:
+            splitted = i.split('_', 1)
+            try:
+                kwds2[splitted[0]][splitted[1]] = kwds[i]
+            except KeyError:
+                raise KeyError('Unknown prefix {0} for parameter {1}'
+                               ''.format(splitted[0], i))
+
+        kwargs = {}
+
+        # First check that the WCS allows the arithmetic operation
+        if compare_wcs is None:
```

Comment:
Actually this is just a question about expectations of the end-user. I remember my frustration with ``matplotlib.pyplot.imshow(interpolation=..)``. That's ugly as hell as there are two values for ``interpolation``: ``None`` and ``'none'``, which do completly different things. So I'm very glad you pointed it out. As the programmer I thought I made it quite clear what is allowed: ``None``, ``False`` and ``callable`` and what happens. But I'm quite close to the code and I see your point about the expectation that if ``False`` is allowed then ``True`` should/could be allowed. But what about the expectation in case of ``'first'`` wouldn't someone then assume that it could also be ``"second"``? 

I think this is something that should be discussed/decided before I change anything. :-)